File: cpp/example_1_spin.cpp
/***************************************************************
* Author: J.P.G. van Dijk
***************************************************************/

// Performance/accuracy settings
//#define MKL					(define at project level)
//#define DOUBLE_PRECISION		(define at project level)

// Includes
#define _USE_MATH_DEFINES
#include <math.h>
#include <iostream>
#include "spine/simulate.h"
#include "spine/fidelity.h"
#include "spine/math.h"
#include "spine/solvers/solver_analytical_xz.h"
#include "spine/systems/system_1_spin.h"

// Namespaces
using namespace std;
using namespace spine::math;

// Function prototypes
bool inHamiltonian(realnum * H, realnum * dt);
void outOperation(complex * U);

// Private global variables/constants
static spine::systems::system_1_spin spin_system = spine::systems::system_1_spin();
static const realnum dt = (realnum) 10e-12;
static int Nsim, nsim;

int main(void)
{
	// Create the system
	spin_system.setLarmorFrequency((realnum)(2 * M_PI * 1e9));
	spin_system.setRabiFrequency((realnum)(2 * M_PI * 10e6));

	// Generate the driving signal
	realnum tpi = (realnum)M_PI / spin_system.getRabiFrequency();
	int Npi = (int)(tpi / dt);

	// Simulate the system
	nsim = 0;
	Nsim = Npi;
#ifdef PLOT
	spin_system.plotSetup(Nsim, 0, tpi);
#endif
	cout << "Simulating..." << endl;
	spine::simulate(spin_system.getDimension(), inHamiltonian, outOperation, spine::solvers::solver_analytical_xz);
	cout << "finished!" << endl;
#ifdef PLOT
	Plot::run();
#endif
}

bool inHamiltonian(realnum * H, realnum * dt)
{
	if (nsim < Nsim)
	{
		// Set the signal at this time instance
		spin_system.setMicrowaveControl(cos(spin_system.getLarmorFrequency() * (nsim * ::dt - 0.5 * ::dt)));

		// Update the Hamiltonian accordingly
		spin_system.updateHamiltonian(H);

		// Provide the current timestep, and continue the simulation
		*dt = ::dt;
		return true;
	}
	return false;
}

void outOperation(complex * U)
{
	// Plot every Nplot-th point
	const unsigned int Nplot = 5;

	// Print the measurement probability
#ifdef PLOT
	spin_system.plotAddU(U, nsim * dt, true);
#endif
	if ((nsim % Nplot) == 0)
	{
		realnum P = abs2(U[0]);
		cout << P << endl;
#ifdef PLOT
		spin_system.plot();
#endif
	}

	// Continue the simulation
	nsim++;
	if (nsim == Nsim)
	{
		// Last point, print the fidelity
		realnum F = spine::fidelity(spin_system.getDimension(), U, (realnum) M_PI, (realnum) 0.0);
		cout << "Fidelity: " << F << endl;
	}
}

File: matlab/example_1_spin.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Author: J.P.G. van Dijk                                                 %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Cleanup
clear all;
close all;
clc;

% Global variables
global nsim;
global Nsim;
global system;
global microwave_control;
global dt;
global probability;

% Create the system
system = spine.systems.system_1_spin();
system.setLarmorFrequency(2 * pi * 1e9);
system.setRabiFrequency(2 * pi * 10e6);

% Generate the driving signal
dt = 10e-12;
tpi = pi / system.getRabiFrequency();
Npi = round(tpi / dt);
t = (0:Npi-1) * dt;
microwave_control = cos(system.getLarmorFrequency() * (t - 0.5*dt));

% Simulate the system
nsim = 0;
Nsim = Npi;
spine.simulate(system.getDimension(), @inHamiltonian, @outOperation, @spine.solvers.solver_analytical_xz);

% Plot the result
figure();
plot(t, probability);

function [run, H, timestep] = inHamiltonian()
    global nsim;
    global Nsim;
    global system;
    global microwave_control;
    global dt;

    if (nsim < Nsim)
        
        % Set the signal at this time instance
        system.setMicrowaveControl(microwave_control(nsim + 1));
        
        % Update the Hamiltonian accordingly
        H = system.updateHamiltonian();
        
        % Provide the current timestep, and continue the simulation
        timestep = dt;
        run = true;
        
    else
        run = false;
        H = [];
        timestep = [];
    end
    
end

function outOperation(U)
    global nsim;
    global Nsim;
    global system;
    global dt;
    global probability;

    % Plot every Nplot-th point
    Nplot = 5;
    
    % Store the measurement probability
    probability(nsim + 1) = abs(U(1,1))^2;
    
    % Show the rotation in the Bloch sphere
    if (mod(nsim, Nplot) == 0)
        system.plot(U, nsim * dt, 2);
        pause(0.001);
    end
    
    % Continue the simulation
    nsim = nsim + 1;
    if (nsim == Nsim)
        
        % Last point, print the fidelity and unitary matrix
        F = spine.fidelity(system.getDimension(), U, pi, 0);
        disp(F);
        disp(U);
        
    end
    
end
